### PR TITLE
feat: evaluator module

### PR DESCRIPTION
**Pull request by khvn26**
- Move segment rule evaluators to flag_engine.segments.evaluator module
- Add missing runtime type checks

### CODE REVIEW

File: flag_engine/segments/evaluator.py [hunk 1]
```diff
@@ -95,4 +108,102 @@ def _traits_match_segment_condition(
...
+
+
+def _matches_trait_value(
+    condition: SegmentConditionModel,
+    trait_value: TraitValue,
+) -> bool:
+    if match_func := MATCH_FUNCS_BY_OPERATOR.get(condition.operator):
+        return match_func(condition.value, trait_value)
+
+    maybe_semver_trait_value: typing.Union[
```
Comment (gagantrivedi) [inline flag_engine/segments/evaluator.py:121]: I think this code is slightly confusing
1. it's very hard to tell what is the difference between match function and operator function(I sort prefer the `._EXCEPTION_OPERATOR_METHODS` approach, at the very least it tells the user that all the functions are match function but some of them are slightly different
2.  I think ` maybe_semver_trait_value` adds nothing but confusion (we can just use the trait value)

**Review by gagantrivedi — CHANGES_REQUESTED** [review]

File: tests/unit/segments/test_segments_evaluator.py [hunk 2]
```diff
@@ -145,25 +157,198 @@ def test_identity_in_segment_percentage_split(
...
+        (constants.LESS_THAN_INCLUSIVE, 2.1, "2.0", False),
+        (constants.NOT_EQUAL, "bar", "baz", True),
+        (constants.NOT_EQUAL, "bar", "bar", False),
+        (constants.NOT_EQUAL, 1, "2", True),
+        (constants.NOT_EQUAL, 1, "1", False),
+        (constants.NOT_EQUAL, True, "False", True),
+        (constants.NOT_EQUAL, False, "True", True),
+        (constants.NOT_EQUAL, False, "False", False),
+        (constants.NOT_EQUAL, True, "True", False),
+        (constants.CONTAINS, "bar", "b", True),
```
[the remark this answers is not in the document]
Comment (khvn26) [reply tests/unit/segments/test_segments_evaluator.py:242]: Added.

File: flag_engine/segments/evaluator.py [hunk 3]
```diff
@@ -95,4 +108,102 @@ def _traits_match_segment_condition(
...
+        return False
+
+    return trait_value % divisor == remainder
+
+
+def _evaluate_in(segment_value: typing.Optional[str], trait_value: TraitValue) -> bool:
+    if segment_value:
+        if isinstance(trait_value, str):
+            return trait_value in segment_value.split(",")
+        elif isinstance(trait_value, int) and not any(
```
[the remark this answers is not in the document]
Comment (khvn26) [reply flag_engine/segments/evaluator.py]: Done.

File: flag_engine/segments/evaluator.py (hunk 1 above)
Comment (khvn26) [reply flag_engine/segments/evaluator.py]: Solved with factoring out the type coercion to a decorator. Now the match function storage is unified.

File: flag_engine/segments/models.py [hunk 4]
```diff
@@ -1,15 +1,10 @@
...
 from pydantic import BaseModel, Field
 
 from flag_engine.features.models import FeatureStateModel
 from flag_engine.segments import constants
 from flag_engine.segments.types import ConditionOperator, RuleType
-from flag_engine.utils.semver import is_semver
-from flag_engine.utils.types import get_casting_function
 
 
 class SegmentConditionModel(BaseModel):
```
Comment (gagantrivedi) [inline flag_engine/segments/models.py:10]: `_EXCEPTION_OPERATOR_METHODS` can be removed now, right?

**Review by gagantrivedi — CHANGES_REQUESTED** [review]

File: flag_engine/segments/models.py (hunk 4 above)
Comment (khvn26) [reply flag_engine/segments/models.py:10]: Good call. Done.

**Review by gagantrivedi — APPROVED** [review]
Great stuff

**Pull request merged**